File: fsharp/luhn/Luhn.fs
module Luhn

open System
open System.Text.RegularExpressions

let valid (number: string) =
    let nows = number.Replace(" ", "")
    Regex.IsMatch(nows, "\d{2,}$") &&
    (nows
       |> Seq.rev
       |> Seq.map (Char.GetNumericValue >> int)
       |> Seq.mapi (fun i n ->
           if i % 2 = 0 then n
           else if n >= 5 then 2 * n - 9
           else 2 * n)
       |> Seq.sum) % 10 = 0

File: csharp/luhn/Luhn.cs
using System.Linq;
using System.Text.RegularExpressions;

public static class Luhn
{
    private static readonly Regex WsRegex = new(@"\s");

    public static bool IsValid(string number)
    {
        var nows = WsRegex.Replace(number, "");

        if (!nows.All(char.IsDigit) || nows.Length < 2)
            return false;
        return nows
            .Reverse()
            .Where(char.IsDigit)
            .Select(c => (int)char.GetNumericValue(c))
            .Select((n, i) => ((i % 2) == 0) ? n : n * 2)
            .Select(n => n > 9 ? n - 9 : n)
            .Sum() % 10 == 0;
    }
}
